Core.lua

```diff
@@ -29,6 +29,10 @@ WarpDeplete.defaultForcesState = {
   currentCount = 0,
   totalCount = 100,
 
+  -- needed to check proper force count if mob that just died would make self.forcesState.completed be marked true
+  preComplete = 0,
+  triggered = false,
```

Comment:
I'm not 100% sure these are needed. Would the following be possible:

- Don't count the last mob we receive through the event if it would push us beyond 100%, instead set `forcesState` to `completed`
- When a mob dies in `OnCombatLogEvent`, check if `forcesState.completed` is `true`, if yes, add that mob to the count
- `triggered` is basically replaced by `count > totalCount`
- All percent displays check for `completed` first and display as full if so, same for the numbers (show `totalCount` if `complete` is `true`, unless unclamping is enabled)

Alternatively, I think we could have an `extraCount` variable, that is only used to count past `totalCount`, that way we keep the original `currentCount` variable clean. In any case, if the last mob we are reported is not accurate, we should simply not add it in my opinion.

Overall, I think I prefer the latter option. Since we can't get the "count past 100%" information from the game's API if I understand it correctly, we would need to sync that information between players if somebody relogs/reloads, and it's possible the number is not accurate a mob doesn't appear in your combat log. If we keep the "count past 100%" in an extra variable, we can simply decide not to show it if we suspect that it could be inaccurate.

Reply:
Not sure if our testing wasn't conclusive enough (we did only run maybe ~30 keys which isn't a big sample size in the grand scheme of things), but some things of note:

- Don't count the last mob we receive through the event if it would push us beyond 100%, instead set forcesState to completed

The reason for `preComplete` is to save `currentCount` right before the mob that pushes us to 100%, so we can retrieve the force of the mob that just died in `WarpDeplete:OnCombatLogEvent(ev)` and accurately add the force onto that and get an accurate total force amount. In our testing, it always did give us proper amounts - though that in and of itself is a small sample size like I said earlier. 

- When a mob dies in OnCombatLogEvent, check if forcesState.completed is true, if yes, add that mob to the count

We couldn't really do this (to get an accurate count) due to the way the code runs in parallel. To my understanding, once a mob dies one of either `WarpDeplete:OnScenarioPOIUpdate(ev)` or `WarpDeplete:OnScenarioCriteriaUpdate(ev)` runs in `Events.lua`. I don't play WoW, so I'm not really familiar with what these do (docs I found couldn't really tell me exactly). `WarpDeplete:OnCombatLogEvent(ev)` ran as well, but once we got to the point of checking `forcesState.completed`, `self:UpdateForces()` always ran and finished first - which in turn would let us not get the mob that died that would push us to 100% (since we wanted an accurate count to the best of our ability). That's the reason for the `triggered` boolean to add an additional check. If we skip the additional boolean, I don't think there'd be a proper way to get a "true" (in a loose sense since combat log can be unreliable) currentCount.

- triggered is basically replaced by count > totalCount

Same reasons as above - `self:UpdateForces()` always ran first before being able to check count > totalCount in `WarpDeplete:OnCombatLogEvent(ev)`. Which is why we needed triggered. 

Though I do think having an `extraCount` is good as well. I just opted to use `currentCount` to try and keep that variable as true as possible (in the sense that even if we go past 100%, that is still "currentCount"). 

We could have just overthought this code as well. I don't doubt that could be a possibility, but based on our testing, that's why we did what we did. If you think we shouldn't bother to try do it since combat log isn't reliable - we can drop this since it isn't supported natively.

Reply:
> The reason for `preComplete` is to save `currentCount` right before the mob that pushes us to 100%, so we can retrieve the force of the mob that just died in `WarpDeplete:OnCombatLogEvent(ev)` and accurately add the force onto that and get an accurate total force amount. In our testing, it always did give us proper amounts - though that in and of itself is a small sample size like I said earlier.

Yeah, I think I understand the purpose of `preComplete`. I'm not suggesting a change in functionality here, just a change in the way the data is represented. After `GetEnemyForcesCount` returns `totalCount`, it basically becomes useless, since it doesn't count above that from what I understand. That means we could do the following (pseudocode):

```lua
local newCurrentCount = self:GetEnemyForcesCount()
if newCurrentCount >= self.forcesState.totalCount then
  -- _Don't_ update self.forcesState.currentCount in this case - we'll keep it and it
  -- will basically act as your new `preComplete` count
  self.forcesState.completed = true

  -- ...
else
  -- Set currentCount normally
end 
```

Then, in the combat log update function:

```lua
if self.forcesState.completed then
  -- We're already triggered in this case, all count we get while completed is extra count
  self.forcesState.extraCount = self.forcesState.extraCount + guidForceCount
else if self.forcesState.currentCount + guidForceCount >= self.forcesState.currentCount then
  -- We just went over, so we can start counting extra count now
  local rest = self.forcesState.totalCount - self.forcesState.currentCount
  self.forcesState.extraCount = guidForceCount - rest

  -- I don't know the order that events come in, it's possible that it makes sense to set
  -- completed here as well
end
```

Is that clearer? If it isn't, maybe it's not a good idea :sweat_smile: But in any case, I think:

- `extraCount` should be separate
- if you want to keep `preComplete`, I think something like `lastKnownAccurateCount` (or something similar) would be a better name, at least `count` should be in the name
- `triggered` should be called something like `countingExtra`, so it's more obvious what it's indicating

Did you make sure you're resetting these 2 values when the state is reset?